### ISSUE TITLE
Investigate Cloudfront caching issue

### ISSUE THREAD

**Issue by MCatherine1994** [bcgov/nr-forests-access-management#1327]
**Describe the task**
DEV and TEST used to have a caching issue, so when there is a frontend code change, it could take up to 1 hour to make frontend back online. We didn't see that issue for PROD in the past, but saw it when release v1.15.0. Need to do some investigation. At the same time, we might need to send out release notification to let people know our service might be available during the prod release. 

**Acceptance Criteria**
- [ ] first
- [ ] second
- [ ] third

**Additional context**
- Add any other context about the task here.
- Or here

**Comment by ianliuwk1019** [comment]
Initial look into the "cloudfront.tf" there is a distribution behavior settings:
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/1174b7b7-75d7-4949-8820-2d5271958904)
The meaning is how long the object stays in cache:
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/09d3f385-cb62-4332-beec-0fc19789505e)
The place where it can be set from the console is from "Behaviors" tab (with click on radio button then Edit):
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/5394688e-fad9-4089-a3e3-470af6c1cafe)
However, the longer is better for the caching and performance for the users.

**Comment by ianliuwk1019** [comment]
Also see an terraform option current is being configured as "**wait_for_deployment = false**".
The explanation from AWS and Terraform is very unclear.
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/6eba92c8-de71-43e6-aa6b-ee9f9c69d42f)
Based on terraform the default value is "true".  It might be worth experimenting on this option.

**Comment by ianliuwk1019** [comment]
Setup Tools space deployment successfully to be used to experiment on caching issue related to deployment.

Tried to do some minor change on frontend or/and backend, and then deploy to Tools space to observe the frontend behavior.
* All deployments (5, nothing related to caching, only frontend + api-gateway terraform configuration, and revert back and forth for the frontend).
* The frontend always behavior normal.
* The frontend can reflect (very simple page content change) frontend change ranging from within 20 seconds to 15 minutes. But generally, pretty **smooth**; just either still displaying old content or new content. **No error** were observed.
* Although seeing some CORS at first deployment (after previous day's fix on CORS, and deploy again), this is different issue (and I seem to be able to fix this from Terraform code and did not experience this for subsequent deployment); I didn't see the error described at Craig's case. I am quite confident, the case Craig described isn't caused by backend (api-gateway or lambda).  (@MCatherine1994  , we may be able to take a note on this to fix what you mentioned that every time when you deployed to PROD, you always need to "Deploy" the Api-Gateway for admin_management)

**Comment by ianliuwk1019** [comment]
3. Our current FAM is coded with "Legacy cache settings". Switching to use **recommended "Cache policy and origin request policy"** option (with AWS '**CachingOptimized**' for S3)to experiment difference.
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/4bac5b31-77b2-4676-b0b8-ae020c59923f)
* Used Terraform (new config) to deploy this option.
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/40b7e991-6761-4a4f-afd7-a1a1d4d968e4)
* The frontend access seems to be smooth after deployment.
* However, it is probably still depending on timing, the error could happen.
   For several deployment tries, I didn't see infinite error popping but with the same error showing up **only once** when it happened and was not continuously popping.
    And there was once after deployment, I **could see infinite error popping again**.
* So even with AWS '**CachingOptimized**' S3 cache option, it does not prevent the error from happening.
* ### But the "**BEST**" outcome so far was, for this option (3): do the deployment, then immediately do "Invalidations". 
   With that, I could observe the browser does not cache file (as I can tell from the resource files requested at browser comparing to S3 bucket origin files, and they match), and the frontend content reflects right away what I changed, and I did not observe a single time for the  infinite error popping. The effect after the "Invalidations" is almost immediate.

**Comment by ianliuwk1019** [comment]
I also tried to observe "**infinite error popping**" (**TypeError: Failed to fetch dynamically imported module**) and tried analyzing it deeper how and what can cause this error.
Still, I can't explain entirely, but I could provide some thoughts after the observation:
* This error happens after new deployment, and I am quite sure it isn't related to backend change (or backend unavailable, failed to return, network connection etc.).
* The error did not just happen at particular page/component, I at least observed it could happen on "profile open" or maybe just when clicking "verify" button.
* The error message might be general, that in some cases frontend might throw out the same error message but could be different scenarios.
* The error is causing from javascript and I am quite sure, in the case we are trying to find out, it happens when browser requested file from javascript but hit **cloudfront caching error** (although the browser indicates the fetching returns http **200**). In this case, I observed that the file being requested is **no longer in S3 origin** (because the new deployment) but **I can't explain why with all the javascript files were no longer in S3, but Cloudfront is able to serve some old stale files but is not able to serve other old stale files, if they have the same expiration.**
* Although the requested file (which causes problem) has http 200 code showing from browser, the response is actually **html** for javascript. This somehow causes frontend to throw error with infinite popping, might be a **bug in Vue**. I kind of remember seeing this in the past but didn't remember it caused infinite popping; maybe situation is different.
* See below error captured from observation:
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/d36a269f-134d-4245-89e4-a9bbe4996525)
  Below two captures are interesting (from the same frontend access time, just different pages). Noted the "**X-Cache**" header, in both cases the requested files no longer exist in S3 (but could still be in Cloudfront cache):
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/c118a394-ebc7-45da-9f96-fa36bbb0b140)
![Image](https://github.com/bcgov/nr-forests-access-management/assets/81595625/011e4ff8-21c1-4941-89df-bf19c521dc21)

**Comment by ianliuwk1019** [comment]
In conclusion:

- Do deployment after office hour.
  So less impact within 1 hour caching and avoid users run into frontned error 
  because of the caching.
  
- Use new [AWS recommended (CachingOptimized) Caching option](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/using-managed-cache-policies.html#managed-cache-caching-optimized) for Cloudfront behavior and **immediately do "Invalidation" after deployment**.
  This seems to have the best result, the frontend reflects its contents right away after "Invalidation"; also probably resolve the frontend infinite error popping(I didn't see it happen once)
  - Need to code it into Terraform (simple, already shown in above example)
  - Need to do invalidation.
    * This is a paid service (but the **first 1,000 invalidation paths is free**, does not matter how many files including in the invalidation path, and we won't need that many invalidations)
	* Can just do "/*" to invalidate all files.
	* There is no option in Terraform to do this "Invalidation" but can do it programmatic way or manually after each deployment.
	  To do "Invalidation" programmatic way, we can use [this example](https://medium.com/@harrysiddhu/5-ways-to-invalidate-cloudfront-cache-82ff6c1a2b17)) described.
	  This example is using AWS Code Pipeline to run but we are using Github Actions so might need to figure out to do that in our Github pipeline and terraform, but should be fine.
	  Basically, do "**Invalidate CloudFront Cache Using S3 Put Event Trigger Lambda**".
   
- With different caching options, the frontend error described by Craig did not go away, it does not happen consistently probably due to the nature of Cloudfront edge location caching behavior for files. This isn't backend issue, it is directly related to Cloudfront caching behavior after deployment,  when fetching javascript files that comes back (with http 200) from Cloudfront with failed content causing some unknown Vue/Vite/Webpack bug. This can probably be resolved if we do above approach.

**Comment by ianliuwk1019** [comment]
Created follow up ticket: #1420 and close this ticket.
